### DISCUSSION TITLE
Custom icon persistence

### DISCUSSION THREAD

**Discussion by Dubzer** [ghostty-org/ghostty#3631] [Feature Requests, Ideas] [answered]
Hi 
I really like icon the customization! 
But what I find strange is that the customized style is only visible when Ghostty is open.

https://github.com/user-attachments/assets/80042c32-017a-4588-9880-d64a30df177f

I was wondering if an option could be added to make the customized icon permanent? I even think it might be a sensible default.

A similar feature is implemented in Telegram for MacOS - ~~Source code~~

https://github.com/user-attachments/assets/2e339bea-0056-48d8-b13a-9c85e476ea89

**Comment by jcollie** [comment]
First, please do not post links to GPL source code. Since Ghostty is MIT licensed we cannot look at GPL code to implement things in Ghostty.
Second, the icon cannot be made persistent because that would invalidate the application's signature on disk. Eventually we will have a way for the user to create an icon file and then they can manually update the icon on disk.

**Reply by Dubzer** [reply]
Oops, sorry about the license! 

I did some more research and found out that it's a well-known trick.
You can write an Extended Attribute to change the icon. And if I understand correctly, this feature is implemented purely at the file system level, so it shouldn't invalidate the signature.
That's basically how changing the icon manually from the "Get Info" works.

Here's an article about Extended attributes: https://eclecticlight.co/2024/05/13/apfs-extended-attributes-revisited/

**Reply by mitchellh** [reply]
If this is true I'm happy to actually add this. Can someone verify or show a proof of concept using `xattr` with the released binary to show it doesn't invalidate the sign/notarization?

**Reply by Dubzer** [reply]
Here's my proof of concept

https://github.com/user-attachments/assets/dcdc6989-9c1e-4fa1-94ce-581f820a46ae

I've learnt a few things while writing it:
1) You actually do need to create a file. It's path should be "./Icon\r". For some reason the icon can't be put directly on the folder. But that doesn't invalidate the signature.
2) I'm not sure which format is used for thumbnails. The easiest way was to generate it is to create a temporary file, let macOS generate a thumbnail, and then just copy it to an attribute on "Icon\r".

Here's the code that I wrote
##### sorry that it's in C#, just haven't gotten around to learning Zig yet lol

<details>
<summary>Code</summary>
If you want to execute this code, you need to have the `Mono.Unix` NuGet package installed in your project

```cs
using System.Diagnostics;
using Mono.Unix.Native;

const string folderPath = "/Users/dubzer/ghostty_test/Ghostty.app";
const string iconPath = "/Users/dubzer/ghostty_test/dog.png";

// I'm not sure what format is used for thumbnails, so we'll let MacOS generate them
Process.Start("/usr/bin/sips", ["-i", iconPath]);
var iconSize = Syscall.getxattr(iconPath, "com.apple.ResourceFork", null!, 0);
var icon = new byte[iconSize];
Syscall.getxattr(iconPath, "com.apple.ResourceFork", icon);

var customIconFlag = new byte[32];
customIconFlag[8] = 0x04;
Syscall.setxattr(folderPath,
    "com.apple.FinderInfo",
    customIconFlag);

var folderIconPath = Path.Combine(folderPath, "Icon\r");

try { File.Delete(folderIconPath); }
catch { }
File.Create(folderIconPath);

Syscall.setxattr(folderIconPath,
    "com.apple.FinderInfo",
    customIconFlag);

Syscall.setxattr(folderIconPath,
    "com.apple.ResourceFork",
    icon);

File.SetAttributes(folderIconPath, File.GetAttributes(folderIconPath) | FileAttributes.Hidden);
```
</details>

Let me know if you would like to see some other checks executed!

**Reply by net** [reply]
Should this be made into an issue? Third-party tools like [Raycast](https://raycast.com/) and [AltTab](https://alt-tab-macos.netlify.app/) use the Finder icon, so users who don't use the dock or native switcher won't benefit from `macos-icon-*` unless it also sets the Finder icon.

Another app that sets the Finder icon dynamically is [Arc](https://arc.net/), so it's definitely possible.

<img width="300" alt="Screenshot 2024-12-31 at 16 22 39" src="https://github.com/user-attachments/assets/f0a7ede9-721b-4948-be27-a383a1aa1891" />
<img width="300" alt="Screenshot 2024-12-31 at 16 23 46" src="https://github.com/user-attachments/assets/cc2f6daa-e455-44a5-8c26-778d3ef0ed6f" />

**Comment by vandorsx** [comment]
Hi, in the meantime I've built a small tool to export a customized icon as a PNG so you can manually apply it in a persistent manner: https://github.com/vandorsx/ghostsmith

update: my tool can now auto apply an icon

**Comment by eli-front** [comment]
I created a rough example proof of concept in zig. I'm relatively new to the language, so please let me know if you see any improvements I should make. The core technologies used are `xattr` & `Rez` https://github.com/eli-front/change-ghostty-icon. @mitchellh If you're willing, I'm interesting in working on / assisting with a PR for this. Based on the `CONTRIBUTING.md`, would the next step be opening an issue?

Edit: Not sure if this would be best implemented in Swift or zig. I will have to learn more about the architecture.

**Comment by sfsam** [comment]
FWIW, there is a macOS shell script that allows the user to set custom icons. It is MIT licensed and available via Homebrew: https://github.com/mklement0/fileicon

In addition, the kitty terminal allows macOS users to use a custom icon by placing it in the kitty config directory: https://sw.kovidgoyal.net/kitty/faq/#i-do-not-like-the-kitty-icon

I like how kitty does it.

**Comment by danielcluff** [comment]
If this feature gets added can we have some icons of the mascot only? The terminal style icons are very busy looking compared to modern ones.

I made my own because Ghostty is so cute
![Screenshot 2025-02-10 at 1 52 15 PM](https://github.com/user-attachments/assets/0b120ee6-4354-435d-b789-02e6d1410ac3)

**Reply by sullicsullic** [reply]
where can I find this cute icon?

**Reply by diego898** [reply]
I'd love this one as well!

**Reply by danielcluff** [reply]
I want to submit this design as an alternate icon. Not sure how to get approval for it.

**Reply by cjahv** [reply]
[222.png.zip](https://github.com/user-attachments/files/20230952/222.png.zip)
This icon was generated using GPT-4o

**Reply by toocomputer** [reply]
Icon-only (official) if anyone's still interested 👻
[Ghostty.icns.zip](https://github.com/user-attachments/files/22407376/Ghostty.icns.zip)

**Comment by sfsam** [comment]
I hacked my copy of Ghostty to look in the config directory for a custom icon ala kitty.
FWIW, it's here: https://github.com/sfsam/ghostty/commit/e7c27cd141e7b46bcb81db4e09dfb3154953f4ed

The difference between this and using [fileicon](https://github.com/mklement0/fileicon) is that it updates `self.appIcon` in `AppDelegate.swift` so Ghostty's About Panel also shows your custom icon:

<img width="412" alt="Screenshot 2025-02-12 at 2 39 32 PM" src="https://github.com/user-attachments/assets/c3841c14-f012-44a8-803c-ec2f6e772c34" />

**Reply by mitchellh** [reply]
I like this a lot! Well done! I think the only thing I'd want to bike shed before being willing to formally accept this is asking what file formats could be accepted here. Right now it's just icns. Does Kitty also require icns?

**Reply by 00-kat** [reply]
If it's possible to have custom icons in GTK as well, it would probably be preferable to accept something that isn't OS-specific, so that people can use the same configuration on both macOS and Linux.  If I remember correctly, a PNG decoder already exists in Ghostty for the Kitty Image Protocol, and PNGs are so widespread, so perhaps PNGs should be accepted too.

> Does Kitty also require icns?

It looks like Kitty supports ICNS and PNG, based on [the following from their FAQ]:
> On macOS and X11 you can put `kitty.app.icns` (macOS only) or `kitty.app.png` in the [kitty configuration directory](https://sw.kovidgoyal.net/kitty/conf/#confloc), and this icon will be applied automatically at startup. On X11, this will set the icon for kitty windows.

[the following from their FAQ]: https://sw.kovidgoyal.net/kitty/faq/#:~:text=On%20macOS%20and%20X11%20you%20can%20put%20kitty.app.icns%20%28macOS%20only%29%20or%20kitty.app.png%20in%20the%20kitty%20configuration%20directory%2C%20and%20this%20icon%20will%20be%20applied%20automatically%20at%20startup.%20On%20X11%2C%20this%20will%20set%20the%20icon%20for%20kitty%20windows.

**Reply by sfsam** [reply]
The [docs](https://developer.apple.com/documentation/appkit/nsworkspace/seticon(_:forfile:options:)) say the `NSWorkspace` `setIcon(_:forFile:options:)` method will accept an arbitrary image:

> The image can be an arbitrary image, with or without transparency. The method automatically scales this image (as needed) to generate the icon representations. The file or folder must exist and be writable by the user.
>
> This method uses the image to set an icon with a size of 512 pixels by 512 pixels. If you specify the [exclude10_4ElementsIconCreationOption](https://developer.apple.com/documentation/appkit/nsworkspace/iconcreationoptions/exclude10_4elementsiconcreationoption) option (not recommended), this method creates an icon that is compatible with the Finder from macOS 10.2 or earlier.

I tested with `icns`, `png`, and `jpg` and all worked even though the filename in all cases is `Ghostty.icns`.

**Reply by coxley** [reply]
@sfsam Forgive me if I've missed it elsewhere in the discussion, but are you open to sharing the icon from your screenshot? It looks awesome! :)

**Reply by sfsam** [reply]
@coxley https://github.com/sfsam/some_icons

**Comment by patrikkaprinay** [comment]
+1

**Reply by 00-kat** [reply]
GitHub has a button for that, y'know.  As a bonus that button also doesn't send an email to everyone subscribed.
![Try press it, maybe.](https://github.com/user-attachments/assets/454fed1a-95e1-4774-af74-a4f92cbbee51)

**Comment by nyngwang** [comment]
Here to link a related issue #3076, where they said that it's currently not possible for `macos-icon-screen-color` setting to light color(s). For finding a workaround, I ended up landing this thread. It's nice to see that plugins made by kind people here seem to be able to cover it.

**Comment by ZuBB** [comment]
a bit off-topic Q: do we have here someone who can create a nice icon based on my prototype?

<img width="589" alt="image" src="https://github.com/user-attachments/assets/0ea1a7e3-59c1-487e-b7f2-e8a19c9eeca4" />

some ideas that might/should be taken into account:
* use ghost emoji as a base
* add '>' and '_' as eyes. those chars may have some glow effect as in original eyes of ghost

At the end would be nice to get something as neat & eye-candy as the following kitty icons




<details>

<summary>good looking kitty icons</summary>
<img width="578" alt="image" src="https://github.com/user-attachments/assets/9e7b5709-b918-4b7b-816a-3830fdd20567" />


https://github.com/DinkDonk/kitty-icon?tab=readme-ov-file#kitty-application-icon
</details>

**Comment by mitchellh** [comment, marked as answer]
This is now done on tip.

**Reply by 00-kat** [reply]
xref #8230.

**Comment by simonseddon** [comment]
Just in case it's useful for anyone else, here's the icon I've made for Ghostty to match the MacOS 26/Tahoe style
<img width="512" height="512" alt="icon_256x256@2x" src="https://github.com/user-attachments/assets/166100cf-734c-45c0-9df9-c2b18465c0a8" />

Or slightly more toned down and on-brand :)
<img width="512" height="512" alt="icon_256x256@2x" src="https://github.com/user-attachments/assets/5bfa174c-5bd4-4e40-bce9-7641f2e428db" />

And my config is like this:
```
macos-custom-icon = ~/Config/ghostty/icon_256x256@2x.png
macos-icon = custom
```

**Reply by ekmag** [reply]
@simonseddon would you be willing to share the original files for these?

**Comment by HendrikPetertje** [comment]
<img width="2000" height="2000" alt="ghostty-rose-pine" src="https://github.com/user-attachments/assets/2fc63e3c-8f16-4674-90ba-2f846549e7fd" />

Rosé Pine for those that like that (inspired by https://github.com/rose-pine/kitty)

**Comment by 00-kat** [comment]
Cool icons y'all've got, but that's unrelated to this discussion (which has also already been closed). If you want to share your custom icons, consider using the `#showcase` channel in the [Ghostty Discord server] instead.

I'm going to lock this discussion since each comment notifies multiple dozen people.

[Ghostty Discord server]: https://discord.gg/ghostty
